week-2-wide-perspective/w02e02-data-manipulation/src/data.js

```diff
@@ -0,0 +1,93 @@
+import axios from 'axios'
+import { Subject } from 'rxjs'
+import { v4 } from 'uuid'
+
+export const STATUS = {
+    loading: 'LOADING',
+    success: 'SUCCESS',
+    fail: 'FAI',
+}
+
+const dummyDelay = request => new Promise((resolve, reject) => {
+   setTimeout(
+       () => {
+           request
+               .then((data) => resolve(data))
+               .catch((error) => reject(error))
+
+    },
+       1000,
+   )
+})
+
+class DataSource {
+    url = ''
+    _status = null
+    data = []
+    onStatusChange
+
+    constructor(uri, onStatusChange) {
+        this.url = `http://localhost:3000${uri}`
+        this.onStatusChange = onStatusChange
+    }
+
+    set status(value) {
+        this.onStatusChange.next(value)
+        this._status = value
+    }
+
+    async get() {
+        this.status = STATUS.loading
+
+        try {
+            const response = await dummyDelay(axios.get(this.url))
+
+            this.data = response.data
+            this.status = STATUS.success
+
+            return response.data
+        } catch (error) {
+            this.status = STATUS.fail
+        }
+    }
+
+    async post(sale) {
+        try {
+            const orderDate = new Date();
+            const response = await axios.post(
+                this.url,
+                {
+                    id: v4(),
+                    orderNumber: `${this.data.length + 1}/${orderDate.getFullYear()}`,
+                    orderDate: orderDate.toJSON(),
+                    sale
+                }
+            )
+
+            this.data.push(response.data)
+            this.status = STATUS.success
+        } catch (error) {}
+    }
+}
+
+
+export const dataSubject = new Subject()
+
+export const registerSubject = (onLoad, onSuccess, onFail) => dataSubject.subscribe({
```

Comment:
Trochę nie rozumiem tego `registerSubject` oraz statusów.

W sensie - rozumiem koncepcję, ale już sama metoda `.subscribe` ma do dyspozycji API:
```javascript
// np. strumień z 1 elementem:
of(1).subscribe( {
   next: (n) => { console.log(n)  },
   error: (err) => { console.log(err) },
   complete: () => { console.log('Stream complete') }
})
```

Użyłem operator `of` w celach pokazowych, żeby mieć strumień, tutaj szczegóły:
https://www.learnrxjs.io/learn-rxjs/operators/creation/of

Więc jeśli przyjmiesz że rozdasz publiczny dostęp do `dataSubject` na zasadzie:

```javascript
export const dataStream = dataSubject.asObservable()
```

natomiast to co przychodzi z `axios` możesz obsługiwać publicznym API dla `Subject`s:
```javascript
import { Subject } from 'rxjs';

const dataSubject = new Subject()

// Zobacz:
dataSubject.next(....)
dataSubject.error(....)
```